packages/storage-plus/src/de.rs

```diff
@@ -0,0 +1,171 @@
+use std::array::TryFromSliceError;
+use std::convert::TryInto;
+
+use cosmwasm_std::{Addr, StdError, StdResult};
+
+use crate::keys::{IntKey, TimestampKey};
+
+pub trait KeyDeserialize {
+    type Output: Sized;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output>;
+}
+
+impl KeyDeserialize for () {
+    type Output = ();
+
+    fn from_slice(_value: &[u8]) -> StdResult<Self::Output> {
+        Ok(())
+    }
+}
+
+impl KeyDeserialize for Vec<u8> {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        Ok(value.to_vec())
+    }
+}
+
+impl KeyDeserialize for &Vec<u8> {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        <Vec<u8>>::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for [u8] {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        <Vec<u8>>::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for &[u8] {
+    type Output = Vec<u8>;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        <Vec<u8>>::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for String {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_utf8(value.to_vec())
+            // FIXME: Add and use StdError utf-8 error From helper
+            .map_err(|err| StdError::generic_err(err.to_string()))
+    }
+}
+
+impl KeyDeserialize for &String {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for str {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for &str {
+    type Output = String;
+
+    fn from_slice(value: &[u8]) -> StdResult<Self::Output> {
+        String::from_slice(value)
+    }
+}
+
+impl KeyDeserialize for Addr {
+    type Output = String;
```

Comment:
Interesting choice the Output isn't Addr.

You handle the utf8 parsing for the String, but since we are reading something from storage, can't we just use `Addr::unchecked`?

Reply:
> Interesting choice the Output isn't Addr.

This is because when we're iterating over these keys, it's usually to return these values from queries. We can change it to `Addr`. In fact, it's the logical choice.

> You handle the utf8 parsing for the String, but since we are reading something from storage, can't we just use `Addr::unchecked`?

Notice this includes an extra step: we first need to convert to `String` using `from_utf8`, and then build an `Addr` using `Addr::unchecked`. This second step is cheap, so, I'm OK with doing this change, for logical consistency.

Reply:
In fact, I think Addr::unchecked should be free for String with a good optimizer